text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,110 @@
+
+### Usage
+
+The following example in Go creates a trace provider that uses resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel api`, `otel sdk`, `otel awsdetector`, and `otel gcpdetector` packages.
+
+```go
+rp := resource.NewMustProvider()
+rp.AddDetectors(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)
+resource := rp.AutoDetect(ctx)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+```
+
+There is no change to how the resource is given to the provider, so no changes are required to the existing spec / implementation.
+
+### Components
+
+#### Detector
+
+The Detector interface will simply return a Resource:
+
+```go
+type Detector interface {
+    Detect(ctx context.Context) (*Resource, error)
+}
+```
+
+If a detector is not able to detect a resource, it must return an empty resource so that the result of each call to `Detect` can be merged.
+
+#### Provider
+
+A Provider will have a function to add detectors & to retrieve the auto-detected reosurce information:
+
+```go
+type Provider struct {
+    AddDetectors(detectors ...Detector) { ... }
+    AutoDetect(ctx context.Context) (*Resource, error) { ... }
+}
+```
+
+### Error Handling
+
+In the case of one or more detectors raising an error, there are two reasonable options:
+
+1. Ignore that detector, and continue with a warning (likely meaning we will continue without expected resource information)
+2. Crash the application (raise a panic)
+
+These options will be provided as separate interfaces to let the user decide how to recover from failure, e.g. `Provider` & `ProviderMust`
+
+## Trade-offs and mitigations
+
+- The resource provider adds a small amount of complexity that may not be necessary if we can't think of any use cases where a user would want to configure different sets of detectors, but omitting it enforces that restriction. We could add a `global.SetResourceProvider(rp)` function similar to the convention for metrics & traces, but there isn't much value in doing this as the resource provider would only be used during initialization.
+- One alternative option to this proposal would be to supply a resource provider directly to the metric or trace provider instead of, or as an additional option to, a static resource. That would result in marginally simpler setup code where the user doesn't need to call `AutoDetect` themselves.
+- In this proposal, no resource detection will happen by default. It may be preferable to make trace / metric providers use the default resource provider if no resource is supplied. This would be in line with the behaviour of the current Java implementation - see [below](#prior-art-and-alternatives)
+- In the case of an error at resource detection time, another alternative would be to start a background thread to retry following some strategy, but it's not clear that there would be much value in doing this, and it would add considerable unnecessary complexity.
```

Comment:
If resource detection is a must, it should just crash an application probably to make the behavior predictable.
Otherwise we may run into same metrics having different resource attributes from time to time / not having resource at all, which will be problematic for aggregation.

Reply:
can always add an option `detector := detectorBuilder.crashOnError(true).build()`

Reply:
I think that's captured by "These options will be provided as separate interfaces to let the user decide how to recover from failure, e.g. `DetectResource` & `MustDetectResource`"